### PR TITLE
[Kernel] Reuse Qwen4Exp HC combine-norm for MTP input

### PR DESCRIPTION
**Pull request by gcanlin**
## Summary

- Reuse the existing Qwen4Exp HC combine kernels for MTP input instead of maintaining an MTP-only add-and-norm kernel.
- Treat `injection_logits=None` as unit injection. Triton specializes directly on the optional `inj_ptr`, so the MTP path performs no injection load or sigmoid.
- Pass `fc_embedding` into the normal decoder as a pending HC block output. The existing combine-norm kernel then fuses the BF16 broadcast add and grouped Gemma RMSNorm.
- Generalize both HC combine variants to the same optional-injection contract, and remove `_forward_mtp_layer` plus the duplicated MTP kernel.

This does not duplicate #54560, which targets the general Qwen4Exp `M=1` SM90 decode path; this change targets the MTP entry residual add and grouped RMSNorm for arbitrary token counts.

## Data flow

```text
target / feedback hidden                 new token
          |                                  |
 pre-norm + fc_hidden                embed + pre-norm
          |                                  |
 residual [T, HC, H]               fc_embedding [T, H]
          |                                  |
          |                         pending block output
          |                         injection = None
          +------------------+---------------+
                             |
                 Qwen4ExpDecoderLayer.forward
                             |
                  attn_hc.combine_and_mix
                             |
             hc_combine_norm[inj_ptr=None]
                    /                    \
       residual + embedding          grouped RMSNorm
                    \                    /
                     +--------+----------+
                              |
                  normal HC mix and decoder
                              |
                     Attention -> MLP
```

## Final kernel specialization

Following review, the final implementation keeps the optional-pointer design while minimizing special-case code:

- Residual and block loads are issued before FP32 conversion and arithmetic.
- Gated and unit-injection paths share the same two-dimensional tiled layout and RMSNorm reduction.
- Both paths use one flattened token-major launch grid, keeping the streams for each token adjacent without an injection-specific grid branch.
- The only unit-injection-specific scheduling is issuing the norm-weight load before the reduction; the gated path keeps the load after the reduction to limit its live range.

## Correctness

```bash
CUDA_VISIBLE_DEVICES=4 python3 -m pytest -q \
  tests/models/qwen4_exp/test_hc_ops.py
```

Result: `8 passed`. The tests cover gated and unit-injection combine paths, including 1, 17, and 2048 tokens.

## Kernel microbenchmark

NVIDIA B200, CUDA 13.0, PyTorch 2.13.0+cu130, BF16, `H=2560`, `HC=4`; CUPTI medians with CUDA graphs and cold L2. All baseline and current callables are captured with `torch.compile(fullgraph=True)`, and both materialized BF16 outputs are checked before timing.

The without-injection baseline is main's separate broadcast add followed by grouped Gemma RMSNorm. The with-injection baseline is main's existing injected HC combine-norm path. Thus each optional-pointer specialization is compared with the corresponding main compiled behavior.

### Without injection

| Tokens | Main compiled (us) | Current (us) | Speedup vs main |
|---:|---:|---:|---:|
| 1 | 3.680 | 2.400 | 1.533x |
| 2 | 4.096 | 2.432 | 1.684x |
| 16 | 4.672 | 2.560 | 1.825x |
| 64 | 6.848 | 2.975 | 2.302x |
| 256 | 15.713 | 4.672 | 3.363x |
| 2048 | 99.968 | 22.175 | 4.508x |
| 8192 | 393.344 | 81.216 | 4.843x |
| 32768 | 1560.477 | 323.968 | 4.817x |

### With injection

| Tokens | Main compiled (us) | Current (us) | Speedup vs main |
|---:|---:|---:|---:|
| 1 | 2.880 | 2.848 | 1.011x |
| 2 | 2.944 | 2.943 | 1.000x |
| 16 | 2.976 | 2.945 | 1.011x |
| 64 | 3.360 | 3.312 | 1.014x |
| 256 | 5.343 | 5.344 | 1.000x |
| 2048 | 25.024 | 23.872 | 1.048x |
| 8192 | 102.400 | 85.152 | 1.203x |
| 32768 | 396.000 | 327.808 | 1.208x |

The injected path is unchanged within measurement noise at small token counts and improves by about 20% at 8192 and 32768 tokens. The unit-injection MTP path remains 1.53x-4.84x faster than the corresponding main compiled sequence.

## End-to-end benchmark

Pending final paired rerun on TP4 B200.

## Model evaluation

Full GSM8K, 1,319 examples, 5-shot, greedy decoding:

```bash
OPENAI_API_KEY=EMPTY PYTHONPATH=. .venv/bin/python -m lm_eval run \
  --model local-chat-completions \
  --model_args model=qwen3.8-flash-next,base_url=http://127.0.0.1:8102/v1/chat/completions,tokenizer_backend=None,tokenized_requests=False,num_concurrent=64,max_gen_toks=1024,max_length=32768,timeout=600 \
  --tasks gsm8k --num_fewshot 5 --apply_chat_template \
  --gen_kwargs temperature=0,max_gen_toks=1024,reasoning_effort=none \
  --seed 42 \
  --output_path /tmp/qwen-mtp-e2e-5k1k-v4/gsm8k-current
```

- Flexible exact match: **0.9666 +/- 0.0049**
- Strict exact match: **0.9659 +/- 0.0050**

Before:

image

After:

image

## AI assistance

OpenAI Codex assisted with implementation, testing, benchmarking, and drafting. The human submitter reviewed the change and is responsible for it.

### CODE REVIEW

**Comment by gcanlin** [comment]
@gau-nernst This PR should be ready now. It's a bit difficult to see the e2e improvement because of the unstable  acceptance.

Would it break the readability? I'm thinking whether this fusion is necessary.

File: vllm/models/qwen4_exp/nvidia/ops/hc.py [hunk 1]
```diff
@@ -215,14 +216,18 @@ def _hc_combine_kernel(
     if launch_pdl:
         tl.extra.cuda.gdc_wait()
 
-    inj = tl.load(inj_ptr + row * stride_inj + offs_hc, mask_hc, other=0.0)
-    block = tl.load(block_ptr + row * stride_block + offs_inner, mask_inner, other=0.0)
+    if HAS_INJECTION:
+        inj = tl.load(inj_ptr + row * stride_inj + offs_hc, mask_hc, other=0.0)
```
Comment (gau-nernst) [inline vllm/models/qwen4_exp/nvidia/ops/hc.py]: i think you can do `if inj_ptr is not None`

File: vllm/models/qwen4_exp/nvidia/ops/hc.py [hunk 2]
```diff
@@ -215,14 +216,18 @@ def _hc_combine_kernel(
...
 
     # Keeping HC as a broadcast dimension is faster here than four separate
     # residual load/store sequences.
-    inj = 2.0 * tl.sigmoid(inj.to(tl.float32) / HC)
-    out = res.to(tl.float32) + block.to(tl.float32)[None, :] * inj[:, None]
+    if HAS_INJECTION:
+        scale = 2.0 * tl.sigmoid(inj.to(tl.float32) / HC)
+    else:
+        scale = tl.full([HC_PAD], 1.0, tl.float32)
+    out = res.to(tl.float32) + block.to(tl.float32)[None, :] * scale[:, None]
```
Comment (gau-nernst) [inline vllm/models/qwen4_exp/nvidia/ops/hc.py]: can we do something like

```python
out = res.to(tl.float32) + block.to(tl.float32)[None, :]
if inj_ptr is not None:
  out = out * (2.0 * tl.sigmoid(inj.to(tl.float32) / HC))
```

File: vllm/models/qwen4_exp/nvidia/ops/hc.py [hunk 3]
```diff
@@ -232,32 +237,37 @@ def _hc_combine_kernel(
...
     assert block_output.shape == (N, hc_dim)
-    assert injection_logits.shape == (N, hc_count)
     assert residual.stride(1) == 1
     assert block_output.stride(1) == 1
-    assert injection_logits.stride(1) == 1
+    if injection_logits is not None:
+        assert injection_logits.shape == (N, hc_count)
+        assert injection_logits.stride(1) == 1
+
+    injection_ptr = injection_logits if injection_logits is not None else residual
```
Comment (gau-nernst) [inline vllm/models/qwen4_exp/nvidia/ops/hc.py]: There is no issue with passing `None` to Triton. It will force a separate specialization anyway.

File: benchmarks/kernels/benchmark_qwen4_exp_mtp_residual_add_norm.py
Comment (gau-nernst) [inline benchmarks/kernels/benchmark_qwen4_exp_mtp_residual_add_norm.py:1]: Please don't commit the microbenchmark file to avoid growing `benchmarks/` folder. You can add it as a snippet in the PR description if you want.

File: vllm/models/qwen4_exp/nvidia/ops/hc.py [hunk 4]
```diff
@@ -278,63 +288,96 @@ def _hc_combine_norm_kernel(
...
-    offs_hc = tl.arange(0, HC_PAD)
-    mask_hc = offs_hc < HC
-    tile_ids = tl.arange(0, NUM_TILES_PAD)
-    offs_inner = tile_ids[:, None] * BLOCK_SIZE + tl.arange(0, BLOCK_SIZE)[None, :]
-    mask_inner = offs_inner < HC_DIM
-    offs = stream * HC_DIM + offs_inner
-    # Shared norm weights repeat across streams; per-branch weights use the
-    # same flattened HC layout as the residual.
-    w_offs = offs_inner if W_SHARED else offs
+    if HAS_INJECTION:
```
Comment (gau-nernst) [inline vllm/models/qwen4_exp/nvidia/ops/hc.py]: Is it possible to avoid large changes in this kernel? Most of the logic should be the same?

**Review by gau-nernst — APPROVED** [review]

**Review by ZJY0516 — APPROVED** [review]

File: vllm/models/qwen4_exp/nvidia/ops/hc.py [hunk 5]
```diff
@@ -343,17 +357,20 @@ def _hc_combine_norm(
...
     assert norm_weight.is_contiguous()
     assert norm_weight.numel() in (hc_dim, DIM)
 
+    stride_injection = injection_logits.stride(0) if injection_logits is not None else 0
+
     out = residual.new_empty(residual.shape)
     y = residual.new_empty(residual.shape)
     BLOCK_SIZE = 512
-    _hc_combine_norm_kernel[(N, hc_count)](
+    _hc_combine_norm_kernel[(N * hc_count,)](
```
Comment (gcanlin) [inline vllm/models/qwen4_exp/nvidia/ops/hc.py:373]: This is the common optimization for w/o injections(can see the PR description to see the benefit). Now the overall change should be minimal I guess.
